src/HydroShocktubeCMA/test_hydro_shocktube_cma.cpp

```diff
@@ -0,0 +1,401 @@
+//==============================================================================
+// TwoMomentRad - a radiation transport library for patch-based AMR codes
+// Copyright 2020 Benjamin Wibking.
+// Released under the MIT license. See LICENSE file included in the GitHub repo.
+//==============================================================================
+/// \file test_hydro_shocktube_CMA.cpp
+/// \brief Defines a test problem for a shock tube with passive scalars using consistent molecular advection (CMA).
+///
+
+#include <cmath>
+#include <string>
+#include <unordered_map>
+
+#include "AMReX_BC_TYPES.H"
+
+#include "ArrayUtil.hpp"
+#include "RadhydroSimulation.hpp"
+#include "fextract.hpp"
+#include "hydro_system.hpp"
+#include "radiation_system.hpp"
+#include "test_hydro_shocktube_cma.hpp"
+#ifdef HAVE_PYTHON
+#include "matplotlibcpp.h"
+#endif
+
+struct ShocktubeProblem {
+};
+
+template <> struct quokka::EOS_Traits<ShocktubeProblem> {
+	static constexpr double gamma = 1.4;
+	static constexpr double mean_molecular_weight = NAN;
+	static constexpr double boltzmann_constant = quokka::boltzmann_constant_cgs;
+};
+
+template <> struct Physics_Traits<ShocktubeProblem> {
+	// cell-centred
+	static constexpr bool is_hydro_enabled = true;
+	static constexpr bool is_chemistry_enabled = false;
+	static constexpr int numMassScalars = 3;		     // number of mass scalars
+	static constexpr int numPassiveScalars = numMassScalars + 0; // number of passive scalars
+	static constexpr bool is_radiation_enabled = false;
+	// face-centred
+	static constexpr bool is_mhd_enabled = false;
+};
+
+// left- and right- side shock states
+constexpr amrex::Real rho_L = 1.0;
+constexpr amrex::Real P_L = 1.0;
+constexpr amrex::Real rho_R = 0.125;
+constexpr amrex::Real P_R = 0.1;
+
+template <> void RadhydroSimulation<ShocktubeProblem>::setInitialConditionsOnGrid(quokka::grid grid_elem)
+{
+	// extract variables required from the geom object
+	amrex::GpuArray<amrex::Real, AMREX_SPACEDIM> dx = grid_elem.dx_;
+	amrex::GpuArray<amrex::Real, AMREX_SPACEDIM> prob_lo = grid_elem.prob_lo_;
+	const amrex::Box &indexRange = grid_elem.indexRange_;
+	const amrex::Array4<double> &state_cc = grid_elem.array_;
+
+	const int ncomp_cc = Physics_Indices<ShocktubeProblem>::nvarTotal_cc;
+	// loop over the grid and set the initial condition
+	amrex::ParallelFor(indexRange, [=] AMREX_GPU_DEVICE(int i, int j, int k) {
+		amrex::Real const x = prob_lo[0] + (i + amrex::Real(0.5)) * dx[0];
+
+		const double vx = 0.0;
+		double rho = NAN;
+		double P = NAN;
+		std::array<Real, 3> specie = {-1.0}; // why can't this loop read in numPassiveScalars
```

Comment:
Why can't this loop read in `numPassiveScalars`?

Reply:
You should be able to. Do you get a compiler error if you try this?

Reply:
You have to use `Physics_Traits<ShocktubeProblem>::numMassScalars`

Reply:
There's no reason. I'm sure I copy-and-pasted the code from somewhere else where I had done one rather than the other. The indices for the hydro variables are consistent between the two (and must be so for the code to work).

Reply:
Are you sure that this is also true for scalars, and `RadSystem` has an index defined for scalars? I don't see anything in `radiation_system.hpp` that is equivalent to `scalar0_index` in `hydro_system.hpp`. So far, whenever I have to use scalars, I have only called `HydroSystem<>` and not `RadSystem<>`